From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Sat, 13 May 2017 20:11:21 -0500
Subject: [PATCH] Add system property to disable book size limits

If anyone comes in with a watchdog crash related to books after this patch
you will not only be publicly shamed but also made an example of.

Disables the security limits on books entirely, allowing plugins AND players
to make books with as much data as they want. Do not use this without
limiting incoming data from packets in some other way.

Removed in 1.17: It was from a different time before books were as jank as they are now. As time has gone on they've only proven to be worse and worse.

diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
index c5d1ba7a1be3f102edcdfdc05fc50b30ef1f775b..ca81c82eef793f3a9a7beaf3aef74743c3f45c1d 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBook.java
@@ -39,6 +39,7 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
     static final int MAX_PAGES = WrittenBookItem.MAX_PAGES; // SPIGOT-6911: Use Minecraft limits
     static final int MAX_PAGE_LENGTH = WrittenBookItem.PAGE_EDIT_LENGTH; // SPIGOT-6911: Use Minecraft limits
     static final int MAX_TITLE_LENGTH = WrittenBookItem.TITLE_MAX_LENGTH; // SPIGOT-6911: Use Minecraft limits
+    private static final boolean OVERRIDE_CHECKS = Boolean.getBoolean("disable.book-limits"); // Paper - Add override
 
     protected String title;
     protected String author;
@@ -235,7 +236,7 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
         if (title == null) {
             this.title = null;
             return true;
-        } else if (title.length() > CraftMetaBook.MAX_TITLE_LENGTH) {
+        } else if (title.length() > CraftMetaBook.MAX_TITLE_LENGTH && !CraftMetaBook.OVERRIDE_CHECKS) { // Paper - Add override
             return false;
         }
 
@@ -437,7 +438,7 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
     String validatePage(String page) {
         if (page == null) {
             page = "";
-        } else if (page.length() > CraftMetaBook.MAX_PAGE_LENGTH) {
+        } else if (page.length() > CraftMetaBook.MAX_PAGE_LENGTH && !CraftMetaBook.OVERRIDE_CHECKS) { // Paper - Add override
             page = page.substring(0, CraftMetaBook.MAX_PAGE_LENGTH);
         }
         return page;
@@ -447,7 +448,7 @@ public class CraftMetaBook extends CraftMetaItem implements BookMeta {
         // asserted: page != null
         if (this.pages == null) {
             this.pages = new ArrayList<String>();
-        } else if (this.pages.size() >= CraftMetaBook.MAX_PAGES) {
+        } else if (this.pages.size() >= CraftMetaBook.MAX_PAGES && !CraftMetaBook.OVERRIDE_CHECKS) { // Paper - Add override
             return;
         }
         this.pages.add(page);
